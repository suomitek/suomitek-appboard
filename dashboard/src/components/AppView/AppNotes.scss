.Terminal {
  overflow-wrap: break-word;
}

.Terminal__Code code {
  white-space: pre-wrap;
}
