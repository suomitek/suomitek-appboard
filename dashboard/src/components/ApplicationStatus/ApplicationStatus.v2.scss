.application-status-pie-chart {
  display: flex;
  justify-content: center;
  text-align: center;
  margin-bottom: -50px;
  &-title {
    margin-top: 1.2rem;
    margin-bottom: 0.6rem;
    font-weight: 600;
  }
  &-label {
    position: relative;
    bottom: 3.6rem;
  }
  &-number {
    margin: 0;
    font-size: 1.5em;
  }
  &-text {
    margin: -5px 0 0 0;
  }
}

.application-status-table {
  text-align: left;
  border-spacing: 1em 0em;
  th {
    text-transform: uppercase;
  }
}
