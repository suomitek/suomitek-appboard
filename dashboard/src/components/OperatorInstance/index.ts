import OperatorInstance from "./OperatorInstance";

export default OperatorInstance;
