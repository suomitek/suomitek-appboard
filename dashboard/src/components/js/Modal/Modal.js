import React from "react";
import PropTypes from "prop-types";

import { get } from "lodash-es";

export const modalSizes = {
  sm: "modal-sm",
  md: "modal-md",
  lg: "modal-lg",
  xl: "modal-xl",
};

const Modal = ({
  title,
  children,
  footer,
  hideCloseButton,
  modalSize,
  onModalClose,
  showModal,
  staticBackdrop,
}) => {
  const onClose = () => onModalClose();

  if (!showModal) {
    return null;
  }

  const size = get(modalSizes, modalSize, modalSizes["md"]);
  const closeOnBackdrop = staticBackdrop ? () => {} : onModalClose;

  return (
    <div className="modal">
      <div
        className={`modal-dialog ${size}`}
        aria-modal={true}
        role="dialog"
        aria-hidden={showModal}
      >
        <div className="modal-content">
          <div className="modal-header">
            {!hideCloseButton && (
              <button
                aria-label="Close the current modal"
                className="close"
                type="button"
                onClick={onClose}
              >
                <clr-icon aria-hidden="true" shape="close" />
              </button>
            )}
            <h3 className="modal-title">{title}</h3>
          </div>
          <div className="modal-body">{children}</div>
          <div className="modal-footer">{footer}</div>
        </div>
      </div>
      <div onClick={closeOnBackdrop} className="modal-backdrop" aria-hidden="true" />
    </div>
  );
};

Modal.defaultProps = {
  modalSize: "md",
  hideCloseButton: false,
  showModal: false,
  staticBackdrop: true,
};

Modal.propTypes = {
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
  footer: PropTypes.node,
  hideCloseButton: PropTypes.bool.isRequired,
  modalSize: PropTypes.string.isRequired,
  onModalClose: PropTypes.func.isRequired,
  showModal: PropTypes.bool.isRequired,
  staticBackdrop: PropTypes.bool,
};

export default Modal;
