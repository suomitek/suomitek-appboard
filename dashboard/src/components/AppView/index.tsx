import * as React from "react";

import { IAppViewProps } from "./AppView";

const AppView = React.lazy(() => import("./AppView"));
const AppViewV2 = React.lazy(() => import("./AppView.v2"));

interface IAppViewSelectorProps extends IAppViewProps {
  UI: string;
}

const AppViewSelector: React.FC<IAppViewSelectorProps> = props => (
  <React.Suspense fallback={null}>
    {props.UI === "clarity" ? <AppViewV2 {...props} /> : <AppView {...props} />}
  </React.Suspense>
);

export default AppViewSelector;
