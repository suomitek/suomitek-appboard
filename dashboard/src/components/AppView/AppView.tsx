import { RouterAction } from "connected-react-router";
import { assignWith, isEqual } from "lodash";
import * as React from "react";
import * as yaml from "yaml";
import { hapi } from "../../shared/hapi/release";

import AccessURLTable from "../../containers/AccessURLTableContainer";
import ApplicationStatus from "../../containers/ApplicationStatusContainer";
import ResourceRef from "../../shared/ResourceRef";
import { IK8sList, IRBACRole, IRelease, IResource } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import AppControls from "./AppControls";
import AppNotes from "./AppNotes";
import AppValues from "./AppValues";
import "./AppView.css";
import ChartInfo from "./ChartInfo";
import ResourceTable from "./ResourceTable";

export interface IAppViewProps {
  cluster: string;
  namespace: string;
  releaseName: string;
  app?: IRelease;
  // TODO(miguel) how to make optional props? I tried adding error? but the container complains
  error: Error | undefined;
  deleteError: Error | undefined;
  getAppWithUpdateInfo: (cluster: string, namespace: string, releaseName: string) => void;
  deleteApp: (
    cluster: string,
    namespace: string,
    releaseName: string,
    purge: boolean,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
}

interface IAppViewState {
  deployRefs: ResourceRef[];
  statefulSetRefs: ResourceRef[];
  daemonSetRefs: ResourceRef[];
  serviceRefs: ResourceRef[];
  ingressRefs: ResourceRef[];
  secretRefs: ResourceRef[];
  otherResources: ResourceRef[];
  manifest: IResource[];
}

export interface IPartialAppViewState {
  deployRefs: ResourceRef[];
  statefulSetRefs: ResourceRef[];
  daemonSetRefs: ResourceRef[];
  serviceRefs: ResourceRef[];
  ingressRefs: ResourceRef[];
  secretRefs: ResourceRef[];
  otherResources: ResourceRef[];
}

const RequiredRBACRoles: { [s: string]: IRBACRole[] } = {
  view: [
    {
      apiGroup: "apps",
      resource: "deployments",
      verbs: ["list", "watch"],
    },
    {
      apiGroup: "apps",
      resource: "services",
      verbs: ["list", "watch"],
    },
  ],
};

class AppView extends React.Component<IAppViewProps, IAppViewState> {
  public state: IAppViewState = {
    manifest: [],
    ingressRefs: [],
    deployRefs: [],
    statefulSetRefs: [],
    daemonSetRefs: [],
    otherResources: [],
    serviceRefs: [],
    secretRefs: [],
  };

  public async componentDidMount() {
    const { releaseName, getAppWithUpdateInfo, cluster, namespace } = this.props;
    getAppWithUpdateInfo(cluster, namespace, releaseName);
  }

  public componentDidUpdate(prevProps: IAppViewProps) {
    const { releaseName, getAppWithUpdateInfo, cluster, namespace, error, app } = this.props;
    if (prevProps.namespace !== namespace || prevProps.cluster !== cluster) {
      getAppWithUpdateInfo(cluster, namespace, releaseName);
      return;
    }
    if (error || !app) {
      return;
    }

    let manifest: IResource[] = yaml
      .parseAllDocuments(app.manifest)
      .map((doc: yaml.ast.Document) => doc.toJSON());
    // Filter out elements in the manifest that does not comply
    // with { kind: foo }
    manifest = manifest.filter(r => r && r.kind);
    if (!isEqual(manifest, this.state.manifest)) {
      this.setState({ manifest });
    } else {
      return;
    }

    // Iterate over the current manifest to populate the initial state
    this.setState(this.parseResources(manifest, app.namespace));
  }

  public render() {
    if (this.props.error) {
      return (
        <ErrorSelector
          error={this.props.error}
          defaultRequiredRBACRoles={RequiredRBACRoles}
          action="view"
          resource={`Application ${this.props.releaseName}`}
          namespace={this.props.namespace}
        />
      );
    }

    return this.props.app && this.props.app.info ? (
      this.appInfo(this.props.app, this.props.app.info)
    ) : (
      <LoadingWrapper />
    );
  }

  public appInfo(app: IRelease, info: hapi.release.IInfo) {
    const { cluster, push } = this.props;
    const {
      serviceRefs,
      ingressRefs,
      deployRefs,
      statefulSetRefs,
      daemonSetRefs,
      secretRefs,
      otherResources,
    } = this.state;
    return (
      <section className="AppView padding-b-big">
        <main>
          <div className="container">
            {this.props.deleteError && (
              <ErrorSelector
                error={this.props.deleteError}
                defaultRequiredRBACRoles={RequiredRBACRoles}
                action="delete"
                resource={`Application ${this.props.releaseName}`}
                namespace={this.props.namespace}
              />
            )}
            <div className="row collapse-b-tablet">
              <div className="col-3">
                <ChartInfo app={app} cluster={cluster} />
              </div>
              <div className="col-9">
                <div className="row padding-t-bigger">
                  <div className="col-4">
                    <ApplicationStatus
                      deployRefs={deployRefs}
                      statefulsetRefs={statefulSetRefs}
                      daemonsetRefs={daemonSetRefs}
                      info={info}
                    />
                  </div>
                  <div className="col-8 text-r">
                    <AppControls
                      cluster={cluster}
                      app={app}
                      deleteApp={this.deleteApp}
                      push={push}
                    />
                  </div>
                </div>
                <AccessURLTable serviceRefs={serviceRefs} ingressRefs={ingressRefs} />
                <AppNotes notes={app.info && app.info.status && app.info.status.notes} />
                <ResourceTable resourceRefs={secretRefs} title="Secrets" />
                <ResourceTable resourceRefs={deployRefs} title="Deployments" />
                <ResourceTable resourceRefs={statefulSetRefs} title="StatefulSets" />
                <ResourceTable resourceRefs={daemonSetRefs} title="DaemonSets" />
                <ResourceTable resourceRefs={serviceRefs} title="Services" />
                <ResourceTable resourceRefs={otherResources} title="Other Resources" />
                <AppValues values={(app.config && app.config.raw) || ""} />
              </div>
            </div>
          </div>
        </main>
      </section>
    );
  }

  private parseResources(
    resources: Array<IResource | IK8sList<IResource, {}>>,
    releaseNamespace: string,
  ): IPartialAppViewState {
    const result: IPartialAppViewState = {
      ingressRefs: [],
      deployRefs: [],
      statefulSetRefs: [],
      daemonSetRefs: [],
      otherResources: [],
      serviceRefs: [],
      secretRefs: [],
    };
    resources.forEach(i => {
      // The item may be a list
      const itemList = i as IK8sList<IResource, {}>;
      if (itemList.items) {
        // If the resource  has a list of items, treat them as a list
        // A List can contain an arbitrary set of resources so we treat them as an
        // additional manifest. We merge the current result with the resources of
        // the List, concatenating items from both.
        assignWith(
          result,
          this.parseResources((i as IK8sList<IResource, {}>).items, releaseNamespace),
          // Merge the list with the current result
          (prev, newArray) => prev.concat(newArray),
        );
      } else {
        const item = i as IResource;
        const resource = { isFetching: true, item };
        switch (i.kind) {
          case "Deployment":
            result.deployRefs.push(
              new ResourceRef(resource.item, this.props.cluster, releaseNamespace),
            );
            break;
          case "StatefulSet":
            result.statefulSetRefs.push(
              new ResourceRef(resource.item, this.props.cluster, releaseNamespace),
            );
            break;
          case "DaemonSet":
            result.daemonSetRefs.push(
              new ResourceRef(resource.item, this.props.cluster, releaseNamespace),
            );
            break;
          case "Service":
            result.serviceRefs.push(
              new ResourceRef(resource.item, this.props.cluster, releaseNamespace),
            );
            break;
          case "Ingress":
            result.ingressRefs.push(
              new ResourceRef(resource.item, this.props.cluster, releaseNamespace),
            );
            break;
          case "Secret":
            result.secretRefs.push(
              new ResourceRef(resource.item, this.props.cluster, releaseNamespace),
            );
            break;
          default:
            result.otherResources.push(
              new ResourceRef(resource.item, this.props.cluster, releaseNamespace),
            );
        }
      }
    });
    return result;
  }

  private deleteApp = (purge: boolean) => {
    return this.props.deleteApp(
      this.props.cluster,
      this.props.namespace,
      this.props.releaseName,
      purge,
    );
  };
}

export default AppView;
