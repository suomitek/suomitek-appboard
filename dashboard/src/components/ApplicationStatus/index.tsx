import * as React from "react";

import { IApplicationStatusProps } from "./ApplicationStatus";

const ApplicationStatus = React.lazy(() => import("./ApplicationStatus"));
const ApplicationStatusV2 = React.lazy(() => import("./ApplicationStatus.v2"));

interface IApplicationStatusSelectorProps extends IApplicationStatusProps {
  UI: string;
}

const ApplicationStatusSelector: React.FC<IApplicationStatusSelectorProps> = props => (
  <React.Suspense fallback={null}>
    {props.UI === "clarity" ? <ApplicationStatusV2 {...props} /> : <ApplicationStatus {...props} />}
  </React.Suspense>
);

export default ApplicationStatusSelector;
