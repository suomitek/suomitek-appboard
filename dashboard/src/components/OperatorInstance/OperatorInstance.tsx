import { RouterAction } from "connected-react-router";
import * as yaml from "js-yaml";
import * as React from "react";

import OperatorNotSupported from "components/OperatorList/OperatorsNotSupported";
import AccessURLTable from "../../containers/AccessURLTableContainer";
import ApplicationStatus from "../../containers/ApplicationStatusContainer";
import placeholder from "../../placeholder.png";
import { fromCRD } from "../../shared/ResourceRef";
import { IClusterServiceVersion, IClusterServiceVersionCRD, IResource } from "../../shared/types";
import { app } from "../../shared/url";
import AppNotes from "../AppView/AppNotes";
import AppValues from "../AppView/AppValues";
import { IPartialAppViewState } from "../AppView/AppView";
import ResourceTable from "../AppView/ResourceTable";
import Card, { CardContent, CardFooter, CardGrid, CardIcon } from "../Card";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";

export interface IOperatorInstanceProps {
  isFetching: boolean;
  cluster: string;
  namespace: string;
  csvName: string;
  crdName: string;
  instanceName: string;
  getResource: (
    namespace: string,
    csvName: string,
    crdName: string,
    resourceName: string,
  ) => Promise<void>;
  deleteResource: (namespace: string, crdName: string, resource: IResource) => Promise<boolean>;
  push: (location: string) => RouterAction;
  errors: {
    fetch?: Error;
    delete?: Error;
    update?: Error;
  };
  resource?: IResource;
  csv?: IClusterServiceVersion;
}

export interface IOperatorInstanceState {
  modalIsOpen: boolean;
  crd?: IClusterServiceVersionCRD;
  resources?: IPartialAppViewState;
}

class OperatorInstance extends React.Component<IOperatorInstanceProps, IOperatorInstanceState> {
  public state: IOperatorInstanceState = {
    modalIsOpen: false,
  };

  public componentDidMount() {
    const { csvName, crdName, instanceName, namespace, getResource } = this.props;
    getResource(namespace, csvName, crdName, instanceName);
  }

  public componentDidUpdate(prevProps: IOperatorInstanceProps) {
    const { csvName, crdName, instanceName, namespace, getResource, resource, csv } = this.props;
    if (prevProps.namespace !== namespace) {
      getResource(namespace, csvName, crdName, instanceName);
      return;
    }
    let crd = this.state.crd;
    if (csv !== prevProps.csv || resource !== prevProps.resource) {
      if (csv && resource) {
        crd = csv.spec.customresourcedefinitions.owned.find(c => c.kind === resource.kind);
        this.setState({ crd });
      }
      if (crd && resource) {
        let result: IPartialAppViewState = {
          ingressRefs: [],
          deployRefs: [],
          statefulSetRefs: [],
          daemonSetRefs: [],
          otherResources: [],
          serviceRefs: [],
          secretRefs: [],
        };
        const ownerRef = { name: resource.metadata.name, kind: resource.kind };
        if (crd.resources) {
          crd.resources?.forEach(r => {
            switch (r.kind) {
              case "Deployment":
                result.deployRefs.push(fromCRD(r, "default", this.props.namespace, ownerRef));
                break;
              case "StatefulSet":
                result.statefulSetRefs.push(fromCRD(r, "default", this.props.namespace, ownerRef));
                break;
              case "DaemonSet":
                result.daemonSetRefs.push(fromCRD(r, "default", this.props.namespace, ownerRef));
                break;
              case "Service":
                result.serviceRefs.push(fromCRD(r, "default", this.props.namespace, ownerRef));
                break;
              case "Ingress":
                result.ingressRefs.push(fromCRD(r, "default", this.props.namespace, ownerRef));
                break;
              case "Secret":
                result.secretRefs.push(fromCRD(r, "default", this.props.namespace, ownerRef));
                break;
              default:
                result.otherResources.push(fromCRD(r, "default", this.props.namespace, ownerRef));
            }
          });
        } else {
          const emptyCRD = { kind: "", name: "", version: "" };
          // The CRD definition doesn't define any service so pull everything
          result = {
            deployRefs: [
              fromCRD(
                { ...emptyCRD, kind: "Deployment" },
                "default",
                this.props.namespace,
                ownerRef,
              ),
            ],
            ingressRefs: [
              fromCRD({ ...emptyCRD, kind: "Ingress" }, "default", this.props.namespace, ownerRef),
            ],
            statefulSetRefs: [
              fromCRD(
                { ...emptyCRD, kind: "StatefulSet" },
                "default",
                this.props.namespace,
                ownerRef,
              ),
            ],
            daemonSetRefs: [
              fromCRD(
                { ...emptyCRD, kind: "DaemonSet" },
                "default",
                this.props.namespace,
                ownerRef,
              ),
            ],
            serviceRefs: [
              fromCRD({ ...emptyCRD, kind: "Service" }, "default", this.props.namespace, ownerRef),
            ],
            secretRefs: [
              fromCRD({ ...emptyCRD, kind: "Secret" }, "default", this.props.namespace, ownerRef),
            ],
            otherResources: [],
          };
        }
        this.setState({ resources: result });
      }
    }
  }

  public render() {
    const {
      isFetching,
      errors,
      resource,
      csv,
      instanceName,
      csvName,
      crdName,
      cluster,
      namespace,
      push,
    } = this.props;
    if (cluster !== "default") {
      return <OperatorNotSupported namespace={namespace} />;
    }
    const { resources } = this.state;
    const onUpdateClick = () =>
      push(app.operatorInstances.update(cluster, namespace, csvName, crdName, instanceName));
    const error = errors.fetch || errors.delete || errors.update;
    return (
      <section className="AppView padding-b-big">
        <main>
          <LoadingWrapper loaded={!isFetching}>
            {error && (
              <ErrorSelector
                error={error}
                action="get"
                resource={`Operator intance ${instanceName}`}
                namespace={namespace}
              />
            )}
            {resource && (
              <div className="row collapse-b-tablet">
                <div className="col-3">{csv && this.renderCSVInfo(csv)}</div>
                <div className="col-9">
                  <div className="row padding-t-bigger">
                    <div className="col-4">
                      {resources && (
                        <ApplicationStatus
                          deployRefs={resources.deployRefs}
                          statefulsetRefs={resources.statefulSetRefs}
                          daemonsetRefs={resources.daemonSetRefs}
                        />
                      )}
                    </div>
                    <div className="col-8 text-r">
                      <button className="button" onClick={onUpdateClick}>
                        Update
                      </button>
                      <button className="button button-danger" onClick={this.openModal}>
                        Delete
                      </button>
                      <ConfirmDialog
                        onConfirm={this.handleDeleteClick}
                        modalIsOpen={this.state.modalIsOpen}
                        loading={this.props.isFetching}
                        closeModal={this.closeModal}
                      />
                    </div>
                  </div>
                  {resources && (
                    <>
                      <AccessURLTable
                        serviceRefs={resources.serviceRefs}
                        ingressRefs={resources.ingressRefs}
                      />
                      {resource.status && (
                        <AppNotes title="Status" notes={yaml.safeDump(resource.status)} />
                      )}
                      <ResourceTable resourceRefs={resources.secretRefs} title="Secrets" />
                      <ResourceTable resourceRefs={resources.deployRefs} title="Deployments" />
                      <ResourceTable
                        resourceRefs={resources.statefulSetRefs}
                        title="StatefulSets"
                      />
                      <ResourceTable resourceRefs={resources.daemonSetRefs} title="DaemonSets" />
                      <ResourceTable resourceRefs={resources.serviceRefs} title="Services" />
                      <ResourceTable
                        resourceRefs={resources.otherResources}
                        title="Other Resources"
                        requestOtherResources={true}
                      />
                      {resource.spec && <AppValues values={yaml.safeDump(resource.spec)} />}
                    </>
                  )}
                </div>
              </div>
            )}
          </LoadingWrapper>
        </main>
      </section>
    );
  }

  private renderCSVInfo = (csv: IClusterServiceVersion) => {
    const { instanceName } = this.props;
    const { crd } = this.state;
    const icon = csv.spec.icon.length
      ? `data:${csv.spec.icon[0].mediatype};base64,${csv.spec.icon[0].base64data}`
      : placeholder;
    return (
      <CardGrid className="ChartInfo">
        <Card>
          <CardIcon icon={icon} />
          <CardContent>
            <h5>{instanceName}</h5>
            <p className="margin-b-reset">{crd?.description}</p>
          </CardContent>
          <CardFooter>
            <div>
              <div>Cluster Service Version: v{csv.spec.version}</div>
              <div>Kind: {crd?.kind}</div>
            </div>
          </CardFooter>
        </Card>
      </CardGrid>
    );
  };

  private openModal = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  private closeModal = async () => {
    this.setState({
      modalIsOpen: false,
    });
  };

  private handleDeleteClick = async () => {
    const { cluster, namespace, resource } = this.props;
    const { crd } = this.state;
    const deleted = await this.props.deleteResource(namespace, crd!.name.split(".")[0], resource!);
    this.closeModal();
    if (deleted) {
      // TODO: The operator routes do not currently include the cluster so ensure
      // it has a default.
      this.props.push(app.apps.list(cluster || "default", namespace));
    }
  };
}

export default OperatorInstance;
