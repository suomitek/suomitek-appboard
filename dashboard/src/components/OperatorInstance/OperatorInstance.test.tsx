import { IPartialAppViewState } from "components/AppView/AppView";
import OperatorNotSupported from "components/OperatorList/OperatorsNotSupported";
import { shallow } from "enzyme";
import * as React from "react";
import Modal from "react-modal";
import AccessURLTable from "../../containers/AccessURLTableContainer";
import ApplicationStatus from "../../containers/ApplicationStatusContainer";
import itBehavesLike from "../../shared/specs";
import { app } from "../../shared/url";
import AppNotes from "../AppView/AppNotes";
import AppValues from "../AppView/AppValues";
import ResourceTable from "../AppView/ResourceTable";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import OperatorInstance, { IOperatorInstanceProps } from "./OperatorInstance";

const defaultProps: IOperatorInstanceProps = {
  isFetching: false,
  cluster: "default",
  namespace: "default",
  csvName: "foo",
  crdName: "foo.kubeapps.com",
  instanceName: "foo-cluster",
  getResource: jest.fn(),
  deleteResource: jest.fn(),
  push: jest.fn(),
  errors: {},
};

itBehavesLike("aLoadingComponent", {
  component: OperatorInstance,
  props: { ...defaultProps, isFetching: true },
});

it("displays an alert if rendered for an additional cluster", () => {
  const props = { ...defaultProps, cluster: "other-cluster" };
  const wrapper = shallow(<OperatorInstance {...props} />);
  expect(wrapper.find(OperatorNotSupported)).toExist();
});

it("gets a resource when loading the component", () => {
  const getResource = jest.fn();
  shallow(<OperatorInstance {...defaultProps} getResource={getResource} />);
  expect(getResource).toHaveBeenCalledWith(
    defaultProps.namespace,
    defaultProps.csvName,
    defaultProps.crdName,
    defaultProps.instanceName,
  );
});

it("gets a resource again if the namespace changes", () => {
  const getResource = jest.fn();
  const wrapper = shallow(<OperatorInstance {...defaultProps} getResource={getResource} />);
  wrapper.setProps({ namespace: "other" });
  expect(getResource).toHaveBeenCalledTimes(2);
});

it("renders an error", () => {
  const wrapper = shallow(
    <OperatorInstance {...defaultProps} errors={{ fetch: new Error("Boom!") }} />,
  );
  expect(wrapper.find(ErrorSelector)).toExist();
  expect(wrapper.find(AppNotes)).not.toExist();
});

describe("renders a resource", () => {
  const csv = {
    metadata: { name: "foo" },
    spec: {
      icon: [{}],
      customresourcedefinitions: {
        owned: [
          {
            name: "foo.kubeapps.com",
            version: "v1alpha1",
            kind: "Foo",
            resources: [{ kind: "Deployment" }],
          },
        ],
      },
    },
  } as any;
  const resource = {
    kind: "Foo",
    metadata: { name: "foo-instance" },
    spec: { test: true },
    status: { alive: true },
  } as any;

  it("renders the resource and CSV info", () => {
    const wrapper = shallow(<OperatorInstance {...defaultProps} />);
    wrapper.setProps({ csv, resource });
    expect(wrapper.find(AppNotes)).toExist();
    expect(wrapper.find(AppValues)).toExist();
    expect(wrapper.find(".ChartInfo")).toExist();
    expect(wrapper.find(ApplicationStatus)).toExist();
    expect(wrapper.find(AccessURLTable)).toExist();
    expect(wrapper.find(ResourceTable)).toExist();
    expect(wrapper).toMatchSnapshot();
  });

  it("deletes the resource", async () => {
    const deleteResource = jest.fn().mockReturnValue(true);
    const push = jest.fn();
    const wrapper = shallow(
      <OperatorInstance {...defaultProps} deleteResource={deleteResource} push={push} />,
    );
    wrapper.setProps({ csv, resource });
    Modal.setAppElement(document.createElement("div"));
    wrapper.find(".button-danger").simulate("click");

    const dialog = wrapper.find(ConfirmDialog);
    expect(dialog.prop("modalIsOpen")).toEqual(true);
    (dialog.prop("onConfirm") as any)();
    expect(deleteResource).toHaveBeenCalledWith(defaultProps.namespace, "foo", resource);
    // wait async calls
    await new Promise(r => r());
    expect(push).toHaveBeenCalledWith(app.apps.list(defaultProps.cluster, defaultProps.namespace));
  });

  it("updates the state with the CRD resources", () => {
    const wrapper = shallow(<OperatorInstance {...defaultProps} />);
    wrapper.setProps({ csv, resource });
    expect(wrapper.state("resources")).toMatchObject({
      deployRefs: [
        {
          apiVersion: "apps/v1",
          filter: {
            metadata: {
              ownerReferences: [
                {
                  kind: "Foo",
                  name: "foo-instance",
                },
              ],
            },
          },
        },
      ],
    });
  });

  it("updates the state with all the resources if the CRD doesn't define any", () => {
    const wrapper = shallow(<OperatorInstance {...defaultProps} />);
    const csvWithoutResource = {
      ...csv,
      spec: {
        ...csv.spec,
        customresourcedefinitions: {
          owned: [
            {
              name: "foo.kubeapps.com",
              version: "v1alpha1",
              kind: "Foo",
            },
          ],
        },
      },
    } as any;
    wrapper.setProps({ csv: csvWithoutResource, resource });
    const resources = wrapper.state("resources") as IPartialAppViewState;
    const resourcesKeys = Object.keys(resources).filter(k => k !== "otherResources");
    resourcesKeys.forEach(k => expect(resources[k].length).toBe(1));
  });

  it("skips AppNotes and AppValues if the resource doesn't have spec or status", () => {
    const wrapper = shallow(<OperatorInstance {...defaultProps} />);
    wrapper.setProps({ csv, resource: { ...resource, spec: undefined, status: undefined } });
    expect(wrapper.find(AppNotes)).not.toExist();
    expect(wrapper.find(AppValues)).not.toExist();
  });
});
