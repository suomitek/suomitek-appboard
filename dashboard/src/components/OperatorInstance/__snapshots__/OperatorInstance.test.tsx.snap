// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<section
  className="AppView padding-b-big"
>
  <main>
    <LoadingWrapper
      loaded={false}
      type={0}
    />
  </main>
</section>
`;

exports[`renders a resource renders the resource and CSV info 1`] = `
<section
  className="AppView padding-b-big"
>
  <main>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <div
        className="row collapse-b-tablet"
      >
        <div
          className="col-3"
        >
          <CardGrid
            className="ChartInfo"
          >
            <Card>
              <CardIcon
                icon="data:undefined;base64,undefined"
              />
              <CardContent>
                <h5>
                  foo-cluster
                </h5>
                <p
                  className="margin-b-reset"
                />
              </CardContent>
              <CardFooter>
                <div>
                  <div>
                    Cluster Service Version: v
                  </div>
                  <div>
                    Kind: 
                    Foo
                  </div>
                </div>
              </CardFooter>
            </Card>
          </CardGrid>
        </div>
        <div
          className="col-9"
        >
          <div
            className="row padding-t-bigger"
          >
            <div
              className="col-4"
            >
              <Connect(ApplicationStatusSelector)
                daemonsetRefs={Array []}
                deployRefs={
                  Array [
                    ResourceRef {
                      "apiVersion": "apps/v1",
                      "cluster": "default",
                      "filter": Object {
                        "metadata": Object {
                          "ownerReferences": Array [
                            Object {
                              "kind": "Foo",
                              "name": "foo-instance",
                            },
                          ],
                        },
                      },
                      "kind": "Deployment",
                      "name": undefined,
                      "namespace": "default",
                    },
                  ]
                }
                statefulsetRefs={Array []}
              />
            </div>
            <div
              className="col-8 text-r"
            >
              <button
                className="button"
                onClick={[Function]}
              >
                Update
              </button>
              <button
                className="button button-danger"
                onClick={[Function]}
              >
                Delete
              </button>
              <ConfirmDialog
                closeModal={[Function]}
                loading={false}
                modalIsOpen={false}
                onConfirm={[Function]}
              />
            </div>
          </div>
          <Connect(AccessURLTable)
            ingressRefs={Array []}
            serviceRefs={Array []}
          />
          <AppNotes
            notes="alive: true
"
            title="Status"
          />
          <ResourceTable
            resourceRefs={Array []}
            title="Secrets"
          />
          <ResourceTable
            resourceRefs={
              Array [
                ResourceRef {
                  "apiVersion": "apps/v1",
                  "cluster": "default",
                  "filter": Object {
                    "metadata": Object {
                      "ownerReferences": Array [
                        Object {
                          "kind": "Foo",
                          "name": "foo-instance",
                        },
                      ],
                    },
                  },
                  "kind": "Deployment",
                  "name": undefined,
                  "namespace": "default",
                },
              ]
            }
            title="Deployments"
          />
          <ResourceTable
            resourceRefs={Array []}
            title="StatefulSets"
          />
          <ResourceTable
            resourceRefs={Array []}
            title="DaemonSets"
          />
          <ResourceTable
            resourceRefs={Array []}
            title="Services"
          />
          <ResourceTable
            requestOtherResources={true}
            resourceRefs={Array []}
            title="Other Resources"
          />
          <AppValues
            values="test: true
"
          />
        </div>
      </div>
    </LoadingWrapper>
  </main>
</section>
`;
